common/tools/dev-tool/src/migrations/migrate-to-esm.ts

```diff
@@ -0,0 +1,260 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT license.
+
+// import { spawn } from "child_process";
+import path from "node:path";
+import { pathExists, readFile, rename, writeFile } from "fs-extra";
+import { createMigration } from "../util/migrations";
+import { Project } from "ts-morph";
+import * as git from "../util/git";
+import semver from "semver";
+import { AzureSdkMetadata, METADATA_KEY } from "../util/resolveProject";
+
+type PackageJsonType = {
+  version: string;
+  main?: string;
+  type?: string;
+  "react-native"?: Record<string, string>;
+  scripts: {
+    "unit-test:node": string;
+    "integration-test:node": string;
+    "unit-test:browser": string;
+    "integration-test:browser": string;
+  };
+  export: {
+    ".": {
+      require: {
+        types: string;
+        default: string;
+      };
+      import: {
+        types: string;
+        default: string;
+      };
+    };
+  };
+  files: string[];
+  devDependencies: Record<string, string>;
+  mocha?: any;
+  [METADATA_KEY]?: AzureSdkMetadata;
+};
+
+export default createMigration(
+  "migrate-to-esm", // unique ID
+  "2023-05-23T21:35:34.784Z", // ISO timestamp that the migration becomes effective
+  "Migrate a package to have type: module in package.json; Add .js extension to imported relative source module; Rename commonjs module to have .cjs extension. etc.", // short description of the migration
+  {
+    async isApplicable(ctx) {
+      // This migration is only applicable to client packages that doesn't have `"type": "module"`.
+      return (
+        ctx.project.packageJson["sdk-type"] === "client" &&
+        ctx.project.packageJson["type"] !== "module"
+      );
+    },
+
+    async validate(_ctx) {
+      /* no-op */
```

Comment:
I probably want to suspend the migration and let users to build and run (and fix) tests before continue.

Reply:
If you want to do that, you can omit the `validate` function altogether and the migration will suspend automatically. It'd be great if we could just try to execute some shim code to validate the ESM conversion, though.

Reply:
Often there are issues that needs manual assist. I think I will leave validation passed so the automated result is commited, then user can fix things on top of it.